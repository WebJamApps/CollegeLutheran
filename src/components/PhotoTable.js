import React, { Component } from 'react';
import MUIDataTable from 'mui-datatables';
import { createMuiTheme, MuiThemeProvider } from '@material-ui/core/styles';
import ReactHtmlParser from 'react-html-parser';
import PropTypes from 'prop-types';
import 'core-js/stable';
import 'regenerator-runtime/runtime';
import { connect } from 'react-redux';
import superagent from 'superagent';
import mapStoreToProps from '../redux/mapStoreToProps';

export class PhotoTable extends Component {
  constructor(props) {
    super(props);
    this.superagent = superagent;
    this.setColumns = this.setColumns.bind(this);
    this.getMuiTheme = this.getMuiTheme.bind(this);
    this.setColumns = this.setColumns.bind(this);
    this.addThumbs = this.addThumbs.bind(this);
    this.state = {
      columns: [],
    };
  }

  componentDidMount() { this.setColumns(); }

  getMuiTheme() { // eslint-disable-line class-methods-use-this
    return createMuiTheme({
      typography: {
        useNextVariants: true,
      },
      overrides: {
        MUIDataTableHeadCell: {
          root: {
            padding: '4px', fontWeight: 'bold', color: 'black', fontSize: '11pt',
          },
        },
        MuiTableRow: { head: { height: '40px' } },
        MuiTableCell: { root: { padding: '4px' } },
        MUIDataTableToolbar: { actions: { display: 'none' }, root: { paddingLeft: 0, minHeight: 'inherit' } },
        MUIDataTable: { responsiveScroll: { maxHeight: '4.3in' } },
        MuiTypography: { h6: { color: 'black', fontWeight: 'bold', fontStyle: 'italic' } },
      },
    });
  }

  setColumns() {
    const columns = [];
    const titles = ['Thumbnail', 'Title', 'Link', 'Type', 'Modify'];
    for (let i = 0; i < titles.length; i += 1) {
      const label = titles[i];// eslint-disable-line security/detect-object-injection
      columns.push({
        name: label.toLowerCase(),
        label,
        options: {
          filter: false,
          sort: true,
          customBodyRender: (value) => (
            <div style={{
              margin: 0, fontSize: '12pt', maxWidth: '200px',
            }}
            >
              { label !== 'Modify' ? ReactHtmlParser(value) : value }
            </div>
          ),
        },
      });
    }
    this.setState({ columns });
  }

  async deletePic(id) { // eslint-disable-next-line no-restricted-globals
    const result = confirm('Deleting picture, are you sure?');// eslint-disable-line no-alert
    if (result) {
      const { auth } = this.props;
      let res;
      try {
        res = await this.superagent.delete(`${process.env.BackendUrl}/book/${id}`)
          .set('Authorization', `Bearer ${auth.token}`).set('Accept', 'application/json');
      } catch (e) {
        console.log(e.message);// eslint-disable-line no-console
        return Promise.resolve(false);
      }
      if (res.status === 200) { window.location.reload(); return Promise.resolve(true); }
      return Promise.resolve(false);
    }
    return Promise.resolve(false);
  }

  addThumbs(arr) { // eslint-disable-line class-methods-use-this
    const newArr = arr;
    for (let i = 0; i < arr.length; i += 1) { // eslint-disable-next-line security/detect-object-injection
      newArr[i].thumbnail = `<img src=${arr[i].url} width="200px"/>`;
      newArr[i].link = `<a href=${arr[i].url} target="_blank">click to view</a>`;// eslint-disable-line security/detect-object-injection
      newArr[i].modify = (<button type="button" onClick={() => this.deletePic(newArr[i]._id)}>Delete Pic</button>);// eslint-disable-line
    }
    return newArr;
  }

  render() {
    const { columns } = this.state;
    const { familyPics, youthPics, otherPics } = this.props;
    let arr = familyPics.concat(youthPics);
    arr = arr.concat(otherPics);
    arr = this.addThumbs(arr);
    return (
      <div className="tourTable">
        <div style={{ maxWidth: '9in', margin: 'auto' }}>
          <MuiThemeProvider theme={this.getMuiTheme()}>
            <MUIDataTable
              options={{
                filterType: 'dropdown',
                pagination: false,
                responsive: 'scrollMaxHeight',
                filter: false,
                download: false,
                search: false,
                print: false,
                viewColumns: false,
                selectableRows: 'none',
                fixedHeader: false,
              }}
              columns={columns}
              data={arr}
              title="All Images"
            />
          </MuiThemeProvider>
        </div>
      </div>
    );
  }
}
PhotoTable.propTypes = {
  // dispatch: PropTypes.func.isRequired,
  // tourUpdated: PropTypes.bool.isRequired,
  auth: PropTypes.shape({ token: PropTypes.string }).isRequired,
  familyPics: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  youthPics: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  otherPics: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
};

export default connect(mapStoreToProps)(PhotoTable);
