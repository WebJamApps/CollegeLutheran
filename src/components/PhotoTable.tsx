import React, { Component } from 'react';
import MUIDataTable from 'mui-datatables';
import { MuiThemeProvider } from '@material-ui/core/styles';
import ReactHtmlParser from 'react-html-parser';
import { HashLink as Link } from 'react-router-hash-link';
import 'core-js/stable';
import 'regenerator-runtime/runtime';
import { connect } from 'react-redux';
import superagent from 'superagent';
import mapStoreToProps from '../redux/mapStoreToProps';
import TableTheme from '../lib/photoTableTheme';

interface Pprops {
  dispatch: (...args: any) => any,
  auth: { token: string },
  familyPics: any[],
  youthPics: any[],
  otherPics: any[],
}
interface Pstate {
  columns: any[]
}
export class PhotoTable extends Component<Pprops, Pstate> {
  superagent: superagent.SuperAgentStatic;

  constructor(props: Readonly<Pprops>) {
    super(props);
    this.superagent = superagent;
    this.setColumns = this.setColumns.bind(this);
    this.setColumns = this.setColumns.bind(this);
    this.handleHideTable = this.handleHideTable.bind(this);
    this.addThumbs = this.addThumbs.bind(this);
    this.state = {
      columns: [],
    };
  }

  componentDidMount() { this.setColumns(); }

  setColumns() {
    const columns: any[] = [];
    const titles = ['Thumbnail', 'Title', 'Caption', 'Link', 'Type', 'Modify'];
    for (let i = 0; i < titles.length; i += 1) {
      const label = titles[i];// eslint-disable-line security/detect-object-injection
      columns.push({
        name: label.toLowerCase(),
        label,
        options: {
          filter: false,
          sort: true,
          customBodyRender: (value: any | null | undefined) => (
            <div style={{
              margin: 0, fontSize: '12pt', maxWidth: '200px',
            }}
            >
              {label !== 'Modify' ? ReactHtmlParser(value) : value}
            </div>
          ),
        },
      });
    }
    this.setState({ columns });
  }

  async deletePic(id: any) { // eslint-disable-next-line no-restricted-globals
    const result = confirm('Deleting picture, are you sure?');// eslint-disable-line no-alert
    if (result) {
      const { auth } = this.props;
      let res: superagent.Response;
      try {
        res = await this.superagent.delete(`${process.env.BackendUrl}/book/${id}`)
          .set('Authorization', `Bearer ${auth.token}`).set('Accept', 'application/json');
      } catch (e) { return `${e.message}`; }
      if (res.status === 200) { window.location.reload(); return Promise.resolve(true); }
      return Promise.resolve(false);
    }
    return Promise.resolve(false);
  }

  editPic(picData: any) {
    const { dispatch } = this.props;
    dispatch({ type: 'EDIT_PIC', picData });
    return true;
  }

  handleHideTable() {
    const { dispatch } = this.props;
    dispatch({ type: 'SHOW_TABLE', showTable: false });
    return true;
  }

  addThumbs(arr: any[]): any[] {
    const newArr = arr;/* eslint-disable security/detect-object-injection */
    for (let i = 0; i < arr.length; i += 1) { // eslint-disable-next-line security/detect-object-injection
      newArr[i].thumbnail = `<img src=${arr[i].url} width="200px"/>`;
      const deletePicId = `deletePic${newArr[i]._id}`;// eslint-disable-line security/detect-object-injection
      const editPicId = `editPic${newArr[i]._id}`;// eslint-disable-line security/detect-object-injection
      newArr[i].link = `<a href=${arr[i].url} target="_blank">click to view</a>`;// eslint-disable-line security/detect-object-injection
      newArr[i].caption = newArr[i].comments === 'showCaption' ? 'display' : 'hide';
      newArr[i].modify = (// eslint-disable-line security/detect-object-injection
        <div>
          <button type="button" id={deletePicId} onClick={() => this.deletePic(newArr[i]._id)}>Delete Pic</button>
          <p>{' '}</p>
          <Link to="/admin/#picsForm">
            <button type="button" id={editPicId} onClick={() => { this.editPic(newArr[i]); this.handleHideTable(); }}>
              Edit Pic
            </button>
          </Link>
        </div>
      );
    }
    return newArr;
  }

  render() {
    const { columns } = this.state;
    const { familyPics, youthPics, otherPics } = this.props;
    let arr: any[] = familyPics.concat(youthPics);
    arr = arr.concat(otherPics); arr = this.addThumbs(arr);
    return (
      <div className="photoTable">
        <div style={{ maxWidth: '9in', margin: 'auto' }}>
          <MuiThemeProvider theme={TableTheme}>
            <MUIDataTable
              options={{
                filterType: 'dropdown',
                pagination: false,
                responsive: 'scrollMaxHeight',
                filter: false,
                download: false,
                search: false,
                print: false,
                viewColumns: false,
                selectableRows: 'none',
                fixedHeader: false,
              }}
              columns={columns}
              data={arr}
              title="All Images"
            />
          </MuiThemeProvider>
        </div>
      </div>
    );
  }
}

export default connect(mapStoreToProps)(PhotoTable);
