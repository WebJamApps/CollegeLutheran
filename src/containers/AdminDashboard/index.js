import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import mapStoreToProps from '../../redux/mapStoreToProps';
import forms from '../../lib/forms';
import AdminController from './AdminController';
import commonUtils from '../../lib/commonUtils';

export class AdminDashboard extends Component {
  constructor(props) {
    super(props);
    this.commonUtils = commonUtils;
    this.controller = new AdminController(this);
    this.state = {
      title: '',
      homePageContent: '',
      forumtitle: '',
      forumurl: '',
      youthName: '',
      youthURL: '',
      forumId: '',
      youthPicsId: '',
      childName: '',
      childURL: '',
      familyPicsId: '',
    };
    this.forms = forms;
    this.onChange = this.onChange.bind(this);
    this.changeHomepage = this.changeHomepage.bind(this);
    this.addForumForm = this.addForumForm.bind(this);
    this.changePicForm = this.changePicForm.bind(this);
    this.deleteFamily = this.deleteFamily.bind(this);
    this.changeFamilyForm = this.changeFamilyForm.bind(this);
    this.deleteForumForm = this.deleteForumForm.bind(this);
  }

  componentDidMount() { this.commonUtils.setTitleAndScroll('Admin Dashboard'); }

  onChange(evt, stateValue) {
    return typeof stateValue === 'string' ? this.setState({ [stateValue]: evt.target.value })
      : this.setState({ [evt.target.id]: evt.target.value });
  }

  deleteFamily() {
    const { familyPicsId } = this.state;
    const { familyPics } = this.props;
    return (
      <form
        id="delete-family"
        style={{
          textAlign: 'left', marginLeft: '4px', width: '100%', maxWidth: '100%',
        }}
      >
        { this.forms.makeDropdown('familyPicsId', '* Select Pic to Delete', familyPicsId, this.onChange, familyPics, '_id', 'title') }
        <div style={{ marginLeft: '60%' }}>
          <p>{' '}</p>
          <button
            onClick={(evt) => this.controller.deleteBookApi(evt, familyPicsId, '/family')}
            type="button"
            disabled={this.controller.validateDeleteBook(familyPicsId)}
          >
            Delete Pic
          </button>
        </div>
      </form>
    );
  }

  changePicForm(picData) {
    const imageUrlValue = this.state[picData.urlId];// eslint-disable-line react/destructuring-assignment
    const imageNameValue = this.state[picData.nameId];// eslint-disable-line react/destructuring-assignment
    return (
      <div className="material-content elevation3" style={{ maxWidth: '320px', margin: 'auto' }}>
        <h4 className="material-header-h4">
          Change
          {' '}
          {picData.title}
          {' '}
          Pictures
        </h4>
        <form>
          <label htmlFor={picData.nameId}>
            Picture Name
            <input id={picData.nameId} value={imageNameValue} onChange={this.onChange} />
          </label>
          <label htmlFor={picData.urlId}>
            Image Address
            <input id={picData.urlId} value={imageUrlValue} onChange={this.onChange} />
          </label>
          <div style={{ marginLeft: '70%' }}>
            <p>{' '}</p>
            <button disabled={picData.disabled()} type="button" id={picData.buttonId} onClick={picData.buttonClick}>Add Pic</button>
          </div>
        </form>
        <hr />
        {picData.deleteSection()}
      </div>
    );
  }

  deleteForumForm(forumId, books) {
    return (
      <form
        id="delete-forum"
        style={{
          textAlign: 'left', margin: 'auto', width: '100%', maxWidth: '100%',
        }}
      >
        { this.forms.makeDropdown('forumId', '* Select Forum to Delete', forumId, this.onChange, books, '_id', 'title') }
        <p>{' '}</p>
        <button
          onClick={(evt) => this.controller.deleteBookApi(evt, forumId, '/news')}
          type="button"
          disabled={this.controller.validateDeleteBook(forumId)}
        >
          Delete Forum
        </button>
      </form>
    );
  }

  addForumForm() {
    const { forumtitle, forumurl, forumId } = this.state;
    const { books } = this.props;
    return (
      <div className="material-content elevation3" style={{ maxWidth: '8in', margin: 'auto' }}>
        <h5>Change Monthly Forum</h5>
        <form
          id="create-forum"
          style={{
            textAlign: 'left', marginLeft: '4px', width: '100%', maxWidth: '100%',
          }}
        >
          {this.forms.makeInput('text', 'Forum Title', false, this.onChange, forumtitle, '90%')}
          {this.forms.makeInput('text', 'Forum URL', false, this.onChange, forumurl, '90%')}
          <div style={{ marginLeft: '70%' }}>
            <p>{' '}</p>
            <button
              type="button"
              id="addForum"
              disabled={this.controller.validateBook(forumtitle, forumurl)}
              onClick={this.controller.addForumAPI}
            >
              Add Forum
            </button>
          </div>
        </form>
        <hr />
        {this.deleteForumForm(forumId, books)}
      </div>
    );
  }

  changeHomepage() {
    const { title, homePageContent } = this.state;
    return (
      <div className="material-content elevation3" style={{ maxWidth: '8in', margin: 'auto' }}>
        <h5>Change Homepage Section</h5>
        <form
          id="create-homepage"
          style={{
            textAlign: 'left', marginLeft: '4px', width: '100%', maxWidth: '100%',
          }}
        >
          {this.forms.makeInput('text', 'Title', false, this.onChange, title, '90%')}
          <label htmlFor="content">
            Content
            <br />
            <textarea id="homePageContent" rows="15" value={homePageContent} style={{ width: '90%' }} onChange={this.onChange} />
          </label>
          <div style={{ marginLeft: '60%' }}>
            <button type="button" id="c-h" disabled={false} onClick={this.controller.createHomeAPI}>Update Homepage</button>
          </div>
        </form>
      </div>
    );
  }

  changeFamilyForm(childName, childURL) {
    return (
      this.changePicForm({
        title: 'Family',
        nameId: 'childName',
        urlId: 'childURL',
        disabled: () => this.controller.validateBook(childName, childURL),
        buttonId: 'addFamilyPic',
        buttonClick: (e) => this.controller.createPicApi(e, {
          title: childName,
          url: childURL,
          comments: childURL,
          type: 'familyPics',
          access: 'CLC',
        }, '/family'),
        deleteSection: this.deleteFamily,
      })
    );
  }

  render() {
    const {
      youthName, youthPicsId, youthURL, childName, childURL,
    } = this.state;
    const { youthPics } = this.props;
    return (
      <div className="page-content">
        <h4 style={{ textAlign: 'center', marginTop: '10px' }}>CLC Admin Dashboard</h4>
        {this.changeHomepage()}
        <p>{' '}</p>
        {this.addForumForm()}
        <p>{' '}</p>
        {this.changePicForm({
          title: 'Youth',
          nameId: 'youthName',
          urlId: 'youthURL',
          disabled: () => this.controller.validateBook(youthName, youthURL),
          buttonId: 'addYouthPic',
          buttonClick: (e) => this.controller.createPicApi(e, {
            title: youthName,
            url: youthURL,
            comments: youthURL,
            type: 'youthPics',
            access: 'CLC',
          }, '/youth'),
          deleteSection: () => this.controller.deleteBookForm('youthPicsId', 'Pic', youthPicsId, youthPics, '/youth'),
        })}
        <p>{' '}</p>
        {this.changeFamilyForm(childName, childURL)}
      </div>
    );
  }
}
AdminDashboard.propTypes = {
  auth: PropTypes.shape({ token: PropTypes.string }).isRequired,
  books: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  youthPics: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  familyPics: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
};
export default withRouter(connect(mapStoreToProps)(AdminDashboard));
