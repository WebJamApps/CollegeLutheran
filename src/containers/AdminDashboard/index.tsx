import React, { Component, Dispatch } from 'react';
import { withRouter, RouteComponentProps } from 'react-router-dom';
import { connect } from 'react-redux';
import mapStoreToProps, { Ibook } from '../../redux/mapStoreToProps';
import forms from '../../lib/forms';
import AdminController from './AdminController';
import commonUtils from '../../lib/commonUtils';
import PTable from '../../components/PhotoTable';

export interface PicData {
  buttonId: string; buttonClick: (e: React.ChangeEvent<EventTarget>) => Promise<string>; title: string; nameId: string;
}
export interface DashboardProps extends RouteComponentProps {
  dispatch: Dispatch<unknown>;
  homeContent: Ibook;
  auth: { token: string };
  books: Record<string, string>[];
  showTable: boolean;
  editPic: Ibook;
  youthPics: Ibook[];
  familyPics: Ibook[];
  otherPics: Ibook[];
  musicPics: Ibook[];
}
type DashboardState = {
  type: string;
  title: string;
  homePageContent: string;
  announcementtitle: string;
  announcementurl: string;
  youthName: string;
  youthURL: string;
  forumId: string;
  showCaption: string;
  firstEdit: boolean;
};
export class AdminDashboard extends Component<DashboardProps, DashboardState> {
  commonUtils: { setTitleAndScroll: (pageTitle: string, width: number) => void; };

  controller: AdminController;

  forms: typeof forms;

  constructor(props: DashboardProps) {
    super(props);
    this.commonUtils = commonUtils;
    this.controller = new AdminController(this);
    this.state = {
      type: '',
      title: props.homeContent.title || '',
      homePageContent: props.homeContent.comments || '',
      announcementtitle: '',
      announcementurl: '',
      youthName: '',
      youthURL: '',
      forumId: '',
      showCaption: '',
      firstEdit: true,
    };
    this.forms = forms;
    this.onChange = this.onChange.bind(this);
    this.onChangeSelect = this.onChangeSelect.bind(this);
    this.checkEdit = this.checkEdit.bind(this);
    this.changeHomepage = this.changeHomepage.bind(this);
    this.changePicForm = this.changePicForm.bind(this);
    this.deleteForumForm = this.deleteForumForm.bind(this);
    this.picButton = this.picButton.bind(this);
    this.handleRadioChange = this.handleRadioChange.bind(this);
    this.resetEditForm = this.resetEditForm.bind(this);
  }

  componentDidMount(): void { this.commonUtils.setTitleAndScroll('Admin Dashboard', window.screen.width); }

  onChange(evt: React.ChangeEvent<HTMLInputElement>, stateValue?: string): string {
    evt.persist();
    this.checkEdit();
    if (typeof stateValue === 'string') {
      this.setState((prevState) => ({ ...prevState, [stateValue]: evt.target.value, firstEdit: false }));
      return stateValue;
    }
    this.setState((prevState) => ({ ...prevState, [evt.target.id]: evt.target.value, firstEdit: false }));
    return evt.target.id;
  }

  onChangeSelect(evt: React.ChangeEvent<HTMLSelectElement>, stateValue?: string): string {
    evt.persist();
    this.checkEdit();
    if (typeof stateValue === 'string') {
      this.setState((prevState) => ({ ...prevState, [stateValue]: evt.target.value }));
      return stateValue;
    }
    this.setState((prevState) => ({ ...prevState, [evt.target.id]: evt.target.value }));
    return evt.target.id;
  }

  checkEdit(): void {
    let {
      youthName, youthURL, type, showCaption,
    } = this.state;
    const { editPic } = this.props;
    if (youthName === '' && editPic.title !== undefined) {
      youthName = editPic.title;
    }
    if (youthURL === '' && editPic.url !== undefined) {
      youthURL = editPic.url;
    }
    if (type === '' && editPic.type !== undefined) {
      type = editPic.type;
    }
    if (showCaption === '' && editPic.comments !== undefined) {
      showCaption = editPic.comments;
    }
    this.setState({
      youthName, youthURL, type, showCaption,
    });
  }

  resetEditForm(evt: React.MouseEvent<HTMLButtonElement>): void {
    evt.preventDefault();
    const { dispatch } = this.props;
    dispatch({ type: 'EDIT_PIC', picData: {} });
    dispatch({ type: 'SHOW_TABLE', showTable: true });
    this.setState({
      youthName: '', youthURL: '', type: '', showCaption: '', firstEdit: true,
    });
  }

  picButton(picData: PicData,
    editPic: { _id: string }, youthName: string, youthURL: string, type: string): JSX.Element {
    const { firstEdit } = this.state;
    return (
      <div style={{ marginLeft: '50%', marginTop: '10px' }}>
        {editPic._id ? (
          <button
            style={{ display: 'relative', marginRight: '20px' }}
            type="button"
            id="cancel-edit-pic"
            onClick={this.resetEditForm}
          >
            Cancel
          </button>
        ) : null}
        <button
          style={{ display: 'relative' }}
          disabled={this.controller.validateBook(youthName, youthURL, type, firstEdit)}
          type="button"
          id={picData.buttonId}
          onClick={
            editPic._id ? this.controller.editPicAPI : picData.buttonClick
          }
        >
          {editPic._id ? 'Edit ' : 'Add '}
          Pic
        </button>
      </div>
    );
  }

  handleRadioChange(evt: { target: { value: string } }): void {
    this.checkEdit();
    this.setState({ showCaption: evt.target.value });
  }

  changePicForm(picData: PicData): JSX.Element['props'] {
    const options = [
      { type: 'youthPics', Category: 'Youth Pics' },
      { type: 'familyPics', Category: 'Family Pics' },
      { type: 'otherPics', Category: 'Other Pics' },
      { type: 'musicPics', Category: 'Music Pics' },
    ];
    const { youthURL, youthName } = this.state;
    let { type, showCaption } = this.state;
    const { editPic } = this.props;
    if (type === '' && editPic.type !== undefined) {
      type = editPic.type;
    }
    if (showCaption === '' && editPic.comments !== undefined) {
      showCaption = editPic.comments;
    }
    return this.controller.changePicDiv(editPic, youthName, youthURL, type, options, showCaption, picData);
  }

  deleteForumForm(forumId: string, options: Record<string, string>[]): JSX.Element {
    const ddParams = {
      htmlFor: 'forumId',
      labelText: '* Select Title to Delete',
      value: forumId,
      onChange: this.onChangeSelect,
      options,
      oValue: '_id',
      dValue: 'title',
    };
    return (
      <form
        id="delete-forum"
        style={{
          textAlign: 'left', margin: 'auto', width: '100%', maxWidth: '100%',
        }}
      >
        {this.forms.makeDataDropdown(ddParams)}
        <p> </p>
        <button
          onClick={(evt) => this.controller.deleteBookApi(evt, forumId, '/news')}
          type="button"
          disabled={this.controller.validateDeleteBook(forumId)}
        >
          Delete Announcement
        </button>
      </form>
    );
  }

  changeHomepage(): JSX.Element {
    const { title, homePageContent } = this.state;
    const inputParams = {
      type: 'text', label: 'Title', isRequired: false, onChange: this.onChange, value: title, width: '90%',
    };
    return (
      <div className="horiz-scroll">
        <div className="material-content elevation3" style={{ width: '850px', margin: '30px auto' }}>
          <h5>Change Homepage Section</h5>
          <form
            id="create-homepage"
            style={{
              textAlign: 'left', marginLeft: '4px', width: '100%', maxWidth: '100%',
            }}
          >
            {this.forms.makeInput(inputParams)}
            <label htmlFor="content">
              Content
              <br />
              {this.controller.editor(homePageContent)}
            </label>
            <div style={{ marginLeft: '60%', marginTop: '10px' }}>
              <button type="button" id="c-h" disabled={false} onClick={this.controller.createHomeAPI}>
                Update Homepage
              </button>
            </div>
          </form>
        </div>
      </div>
    );
  }

  changeYouthForm(): string {
    const {
      youthName, youthURL, type, showCaption,
    } = this.state;
    const postBody = {
      title: youthName,
      url: youthURL,
      comments: showCaption,
      type,
      access: 'CLC',
    };
    return this.changePicForm({
      buttonId: 'addYouthPic',
      buttonClick: (e: React.ChangeEvent<EventTarget>) => this.controller.createPicApi(e, postBody, '/admin'),
      title: '',
      nameId: 'youthName',
    });
  }

  render(): JSX.Element {
    const {
      showTable, auth, dispatch, youthPics, familyPics, otherPics, musicPics,
    } = this.props;
    return (
      <div className="page-content">
        <h4 style={{ textAlign: 'center', marginTop: '10px' }}>
          CLC Admin Dashboard
        </h4>
        {this.changeHomepage()}
        {this.controller.addForumForm()}
        {this.changeYouthForm()}
        {showTable ? (
          <PTable
            auth={auth}
            dispatch={dispatch}
            youthPics={youthPics}
            familyPics={familyPics}
            otherPics={otherPics}
            musicPics={musicPics}
          />
        ) : null}
      </div>
    );
  }
}

export default withRouter(connect(mapStoreToProps, null)(AdminDashboard));
