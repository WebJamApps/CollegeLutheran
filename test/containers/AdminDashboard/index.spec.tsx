/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { shallow } from 'enzyme';
import { AdminDashboard } from '../../../src/containers/AdminDashboard';
import { PhotoTable } from '../../../src/components/PhotoTable';

describe('Dashboard Container', () => {
  let props: any, wrapper: any;
  const history: any = {};
  const location: any = {};
  const match: any = {};
  const editPic:any = {
    title: '', _id: '', type: '', created_at: '',
  };
  beforeEach(() => {
    props = {
      auth: { token: 'token' },
      books: [{ _id: '123' }],
      youthPics: [{ _id: '456' }],
      familyPics: [{ _id: '789' }],
      otherPics: [{ _id: '999' }],
      homeContent: { title: '', comments: '' },
    };
    wrapper = shallow<AdminDashboard>(<AdminDashboard
      dispatch={(fun) => fun}
      auth={props.auth}
      books={props.books}
      homeContent={props.homeContent}
      showTable
      editPic={editPic}
      history={history}
      location={location}
      match={match}
      youthPics={[]}
      familyPics={[]}
      otherPics={[]}
      musicPics={[]}
    />);
  });
  it('renders correctly', () => { expect(wrapper).toMatchSnapshot(); });
  it('handles click to delete a forum', () => {
    wrapper.instance().controller.deleteBookApi = jest.fn();
    wrapper.update();
    const form = wrapper.instance().deleteForumForm('id', []);
    const f = shallow(form);
    f.find('button').simulate('click');
    expect(wrapper.instance().controller.deleteBookApi).toHaveBeenCalled();
  });
  it('handles click to create youth pic', () => {
    wrapper.instance().controller.createPicApi = jest.fn();
    wrapper.update();
    wrapper.instance().setState({ youthName: 'name', youthURL: 'url', type: 'youthPics' });
    const form = wrapper.instance().changeYouthForm();
    const f = shallow(form);
    const createForm = f.find('form').get(0);
    const button = shallow(createForm);
    button.find('button#addYouthPic').simulate('click');
    expect(wrapper.instance().controller.createPicApi).toHaveBeenCalled();
  });
  it('uses the stateValue on change', () => {
    const result = wrapper.instance().onChange({ persist: jest.fn(), target: {} }, 'stateValue');
    expect(result).toBe('stateValue');
  });
  it('uses the stateValue on selectChange', () => {
    const result = wrapper.instance().onChangeSelect({ persist: jest.fn(), target: {} }, 'stateValue');
    expect(result).toBe('stateValue');
  });
  it('uses the event target id on change', () => {
    const result = wrapper.instance().onChange({ persist: jest.fn(), target: { id: 'youthPicsId', value: '456' } });
    expect(result).toBe('youthPicsId');
  });
  it('uses the event target id on changeSelect', () => {
    const result = wrapper.instance().onChangeSelect({ persist: jest.fn(), target: { id: 'youthPicsId', value: '456' } });
    expect(result).toBe('youthPicsId');
  });
  it('renders with edit pic form', () => {
    const editPic2:any = {
      title: '', url: 'url', type: 'otherPics', _id: '123', created_at: '',
    };
    const wrapper2 = shallow<AdminDashboard>(<AdminDashboard
      dispatch={(fun) => fun}
      showTable
      auth={props.auth}
      books={props.books}
      familyPics={props.familyPics}
      otherPics={props.otherPics}
      homeContent={props.homeContent}
      youthPics={props.youthPics}
      musicPics={props.musicPics}
      editPic={editPic2}
      history={history}
      location={location}
      match={match}
    />);
    wrapper2.instance().setState = jest.fn();
    const cpf = wrapper2.instance().changePicForm({
      buttonId: '', buttonClick: jest.fn(), title: '', nameId: '',
    });
    const d = shallow(cpf);
    const h = d.find('h4').text();
    expect(h).toBe('Edit Pictures');
  });
  it('checks for edit data to set state for edit pictures', () => {
    const editPic2:any = {
      title: 'title', url: 'url', type: 'otherPics', _id: '123', comments: 'showCaption', created_at: '',
    };
    const wrapper2 = shallow<AdminDashboard>(<AdminDashboard
      dispatch={(fun) => fun}
      showTable
      auth={props.auth}
      books={props.books}
      youthPics={props.youthPics}
      familyPics={props.familyPics}
      otherPics={props.otherPics}
      musicPics={props.musicPics}
      homeContent={props.homeContent}
      editPic={editPic2}
      history={history}
      location={location}
      match={match}
    />);
    wrapper2.instance().setState = jest.fn();
    wrapper2.update();
    wrapper2.instance().checkEdit();
    expect(wrapper2.instance().setState).toHaveBeenCalled();
  });
  it('checks for edit data to set state when not edit pictures', () => {
    const editPic2: any = {};
    const wrapper2 = shallow<AdminDashboard>(<AdminDashboard
      dispatch={(fun) => fun}
      showTable
      auth={props.auth}
      books={props.books}
      youthPics={props.youthPics}
      familyPics={props.familyPics}
      otherPics={props.otherPics}
      musicPics={props.musicPics}
      homeContent={props.homeContent}
      editPic={editPic2}
      history={history}
      location={location}
      match={match}
    />);
    wrapper2.instance().setState = jest.fn();
    wrapper2.update();
    wrapper2.instance().checkEdit();
    expect(wrapper2.instance().setState).toHaveBeenCalled();
  });
  it('sets state from a radio button change', () => {
    // eslint-disable-next-line jest/no-conditional-expect
    wrapper.instance().setState = jest.fn((obj) => { if (obj.showCaption) expect(obj.showCaption).toBe('showCaption'); });
    wrapper.update();
    wrapper.instance().handleRadioChange({ target: { value: 'showCaption' } });
  });
  it('resets the edit pic form', () => {
    wrapper.instance().setState = jest.fn((obj) => expect(obj.youthName).toBe(''));
    wrapper.update();
    wrapper.instance().resetEditForm({ preventDefault: () => { } });
  });
  it('doesnt show PTable when showTable is true', () => {
    const editPic2:any = {
      title: 'title', url: 'url', type: 'otherPics', _id: '123', comments: 'showCaption', created_at: '',
    };
    const wrapper2 = shallow<AdminDashboard>(<AdminDashboard
      dispatch={(fun) => fun}
      auth={props.auth}
      books={props.books}
      youthPics={props.youthPics}
      familyPics={props.familyPics}
      musicPics={props.musicPics}
      otherPics={props.otherPics}
      homeContent={props.homeContent || ''}
      editPic={editPic2}
      showTable={false}
      history={history}
      location={location}
      match={match}
    />);
    expect(wrapper2.find(PhotoTable).exists()).toBe(false);
  });
});
