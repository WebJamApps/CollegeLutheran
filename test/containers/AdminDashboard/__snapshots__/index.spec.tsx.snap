// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Container renders AdminDashboard 1`] = `
<div
  className="page-content"
>
  <h4
    style={
      {
        "marginTop": "10px",
        "textAlign": "center",
      }
    }
  >
    CLC Admin Dashboard
  </h4>
  <div
    className="horiz-scroll"
  >
    <div
      className="material-content elevation3"
      style={
        {
          "margin": "30px auto",
          "width": "850px",
        }
      }
    >
      <h5>
        Change Homepage Section
      </h5>
      <form
        id="create-homepage"
        style={
          {
            "marginLeft": "4px",
            "maxWidth": "100%",
            "textAlign": "left",
            "width": "100%",
          }
        }
      >
        <label
          className="inquiryLabel"
          htmlFor="title"
        >
          
          Title
          <br />
          <input
            id="title"
            name="title"
            onChange={[Function]}
            required={false}
            style={
              {
                "minWidth": "inherit",
                "paddingLeft": 0,
                "width": "90%",
              }
            }
            type="text"
            value=""
          />
        </label>
        <p>
          Content
        </p>
        <textarea
          id="tiny-react_95361480511483228800000"
          style={
            {
              "visibility": "hidden",
            }
          }
        />
      </form>
    </div>
  </div>
</div>
`;
